### PR TITLE
Bug/#781 create results dir

### LINKED ISSUE #781: Create results-dir
Bug reported via Slack.

- Flank should create results-dir if it doesn't exist
- Let's add a test to verify Flank creates results-dir when the dir doesn't exist

> pavelnazimok
>
> Have this in config:
> 
> ```  
> results-bucket: sdk_test_results
>   results-dir: test_results-6282-2020-05-13_10:34:59
> ```
>
> 
> But Flank gives me this error:
> `'test_results-6282-2020-05-13_10:34:59' from results-dir doesn't exist`
> Using last release
> The same config worked without error on pre v20 snapshot

### PR DESCRIPTION
**Pull request by adamfilipow92**
Fixes #781 

## Checklist

- [X] Documented
- [X] Unit tested
- [X] release_notes.md updated

### CODE REVIEW

File: test_runner/src/main/kotlin/ftl/args/AndroidArgs.kt [hunk 1]
```diff
@@ -55,19 +55,22 @@ class AndroidArgs(
     private val androidGcloud = androidGcloudYml.gcloud
     var appApk = (cli?.app ?: androidGcloud.app ?: throw FlankFatalError("app is not set")).processFilePath("from app")
     var testApk = (cli?.test ?: androidGcloud.test)?.processFilePath("from test")
-    val additionalApks = (cli?.additionalApks ?: androidGcloud.additionalApks).map { it.processFilePath("from additional-apks") }
+    val additionalApks =
```
Comment (jan-gogo) [inline test_runner/src/main/kotlin/ftl/args/AndroidArgs.kt:58]: Pls let's try to avoid autoformat changes

Comment (adamfilipow92) [reply test_runner/src/main/kotlin/ftl/args/AndroidArgs.kt]: Right! I set auto line breaking to 200 chars but this changes was made before. It shouldn't happen any more.

File: test_runner/src/test/kotlin/ftl/args/AndroidArgsTest.kt [hunk 2]
```diff
@@ -1252,6 +1246,32 @@ AndroidArgs
...
+     * really not exist on disk (when issue come back and directory exists, assert not throws),
+     * second for checking directory not created after load args
+     * it's should be enough
+     */
+    @Test
+    fun `should not throw if directory not exists`() {
+        val resultsDir = UUID.randomUUID().toString()
+        val directoryPath = Paths.get(resultsDir)
+        if (Files.exists(directoryPath)) {
+            Assert.fail("Test directory ($resultsDir) shouldn't exists! It's remote directory.")
```
Comment (bootstraponline) [inline test_runner/src/test/kotlin/ftl/args/AndroidArgsTest.kt:1261]: > It's a remote directory.

File: test_runner/src/test/kotlin/ftl/args/AndroidArgsTest.kt [hunk 3]
```diff
@@ -1252,6 +1246,32 @@ AndroidArgs
...
+        }
+        val yaml = """
+        gcloud:
+          app: $appApk
+          test: $testApk
+          results-dir: $resultsDir
+        """.trimIndent()
+        AndroidArgs.load(yaml)
+        if (Files.exists(directoryPath)) {
+            Assert.fail("Test directory ($resultsDir) shouldn't be created! It's remote directory on cloud!")
```
Comment (bootstraponline) [inline test_runner/src/test/kotlin/ftl/args/AndroidArgsTest.kt:1271]: > It's a remote directory on the cloud!

File: test_runner/src/test/kotlin/ftl/args/AndroidArgsTest.kt [hunk 4]
```diff
@@ -1252,6 +1246,32 @@ AndroidArgs
...
+        }
+    }
 }
 
-private fun AndroidArgs.Companion.load(yamlData: String, cli: AndroidRunCommand? = null): AndroidArgs = load(StringReader(yamlData), cli)
+private fun AndroidArgs.Companion.load(yamlData: String, cli: AndroidRunCommand? = null): AndroidArgs =
+    load(StringReader(yamlData), cli)
```
Comment (bootstraponline) [inline test_runner/src/test/kotlin/ftl/args/AndroidArgsTest.kt:1277]: format change? 🙂

File: release_notes.md [hunk 5]
```diff
@@ -1,5 +1,6 @@
 ## next (unreleased)
 
+- [#781](https://github.com/Flank/flank/pull/781) Create results-dir. ([adamfilipow92](https://github.com/adamfilipow92))
```
Comment (bootstraponline) [inline release_notes.md:3]: The bug is, we should not call `assertFileExists` on `gcloud.resultsDir`. I'd update the release notes to say something like:

>  [#781](https://github.com/Flank/flank/pull/781) Remove local exists check on cloud results-dir. Fixes crash when resultsDir is set by the user. ([adamfilipow92](https://github.com/adamfilipow92))

File: test_runner/src/test/kotlin/ftl/args/AndroidArgsTest.kt [hunk 6]
```diff
@@ -1252,6 +1246,32 @@ AndroidArgs
...
     }
+
+    /**
+     * Double directory check for regressive issue protect, first to check directory
+     * really not exist on disk (when issue come back and directory exists, assert not throws),
+     * second for checking directory not created after load args
+     * it's should be enough
+     */
+    @Test
+    fun `should not throw if directory not exists`() {
```
Comment (bootstraponline) [inline test_runner/src/test/kotlin/ftl/args/AndroidArgsTest.kt:1257]: Comments should usually describe why we're doing something, not what we're doing.

> Usually comments are useful when they explain why some code exists, and should not be explaining what some code is doing. 
>
> https://github.com/google/eng-practices/blob/master/review/reviewer/looking-for.md#comments


What do you think about this?
```kotlin
   @Test
   fun `results-dir (cloud directory) should not throw if it doesn't exist locally`
```

File: test_runner/src/test/kotlin/ftl/fixtures/test_app_cases/flank-single-result-dir.yml [hunk 7]
```diff
@@ -0,0 +1,7 @@
+gcloud:
+  app: ./src/test/kotlin/ftl/fixtures/tmp/apk/app-debug.apk
```
Comment (bootstraponline) [inline test_runner/src/test/kotlin/ftl/fixtures/test_app_cases/flank-single-result-dir.yml:2]: Is this used anywhere? I think we inlined the example

Comment (adamfilipow92) [reply test_runner/src/test/kotlin/ftl/fixtures/test_app_cases/flank-single-result-dir.yml:2]: It's only used for my internal tests of this issue. I removed it from repository.

File: test_runner/src/test/kotlin/ftl/args/AndroidArgsTest.kt (hunk 6 above)
Comment (adamfilipow92) [reply test_runner/src/test/kotlin/ftl/args/AndroidArgsTest.kt:1257]: Thanks for this feedback! I change test name and remove my comment. :)

File: release_notes.md (hunk 5 above)
Comment (adamfilipow92) [reply release_notes.md:3]: Changed!

File: test_runner/src/test/kotlin/ftl/args/AndroidArgsTest.kt (hunk 4 above)
Comment (adamfilipow92) [reply test_runner/src/test/kotlin/ftl/args/AndroidArgsTest.kt:1277]: Sadly yes. I should check all files modified in this pr not only one. I revert changes. In future I will pay more attention for that.

File: test_runner/src/test/kotlin/ftl/args/AndroidArgsTest.kt (hunk 2 above)
Comment (adamfilipow92) [reply test_runner/src/test/kotlin/ftl/args/AndroidArgsTest.kt:1261]: Changed

File: test_runner/src/test/kotlin/ftl/args/AndroidArgsTest.kt (hunk 3 above)
Comment (adamfilipow92) [reply test_runner/src/test/kotlin/ftl/args/AndroidArgsTest.kt:1271]: Changed

**Pull request merged**